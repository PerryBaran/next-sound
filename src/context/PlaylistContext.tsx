'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
