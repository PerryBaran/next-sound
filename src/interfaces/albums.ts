interface Songs {
  name: string
  url: string
  id: string
  createdAt: string
}

interface Users {
  name: string
  id: string
  createdAt: string
}

export default interface Albums {
  name: string
  url: string
  id: string
  User: Users
  Songs: Songs[]
  UserId: string
  createdAt: string
}
