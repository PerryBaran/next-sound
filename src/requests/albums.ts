import {
  postRequest,
  getRequest,
  patchRequest,
  deleteRequest,
  getByIdRequest
} from "./helpers"

export async function postAlbums(data: { name: string; image?: File }) {
  const response = await postRequest("albums", data)
  return response
}

export async function getAlbums(query?: {
  name?: string
  exact?: boolean
  limit?: number
}) {
  const response = await getRequest("albums", query)
  return response
}

export async function getAlbumById(id: string) {
  const response = await getByIdRequest("albums", id)
  return response
}

export async function patchAlbum(
  id: string,
  data: { name?: string; image?: File }
) {
  const response = await patchRequest("albums", id, data)
  return response
}

export async function deleteAlbum(id: string) {
  const response = await deleteRequest("albums", id)
  return response
}
