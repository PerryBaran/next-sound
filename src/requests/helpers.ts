import Cookie from 'js-cookie'
import axios from 'axios'

const instance = axios.create({
  withCredentials: true,
  baseURL: 'https://soundcloud-clone-api.onrender.com'
})

function config() {
  return {
    headers: {
      userToken: Cookie.get('userToken')
    }
  }
}

type Model = 'users' | 'users/login' | 'users/signup' | 'albums' | 'string'

function createForm(model: string, data: any) {
  const formData = new FormData()
  switch (model) {
    case 'songs': {
      if (data.name) {
        formData.append('name', data.name)
      }
      if (data.audio) {
        formData.append('audio', data.audio)
      }
      if (data.position !== undefined) {
        formData.append('position', data.position)
      }
      if (data.AlbumId) {
        formData.append('AlbumId', data.AlbumId)
      }
      break
    }
    case 'albums': {
      if (data.name) {
        formData.append('name', data.name)
      }
      if (data.image) {
        formData.append('image', data.image)
      }
      break
    }
    default: {
      return data
    }
  }
  return formData
}

function error(err: any | { response: { data: { message: string } } }) {
  if (err.response?.data?.message) {
    throw new Error(err.response.data.message)
  }
  throw new Error('Unexpected Error')
}

export async function postRequest(model: Model, data: any) {
  const formData = createForm(model, data)

  try {
    const response = await instance.post(`/${model}`, formData, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function getRequest(
  model: Model,
  query:
    | {
        name?: string
        exact?: boolean
        limit?: number | string
      }
    | undefined
) {
  let endpoint = `/${model}`
  const queryArray = []
  if (query) {
    if (query.name) {
      queryArray.push(`name=${query.name}`)
      if (query.exact) {
        queryArray.push('exact=true')
      }
    }
    if (query.limit) {
      queryArray.push(`limit=${query.limit}`)
    }
    if (queryArray.length > 0) {
      const query = queryArray.join('&')
      endpoint = `${endpoint}?${query}`
    }
  }

  try {
    const response = await instance.get(endpoint, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function getByIdRequest(model: Model, id: string | number) {
  try {
    const response = await instance.get(`/${model}/${id}`, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function patchRequest(
  model: Model,
  id: string | number,
  data: any
) {
  const formData = createForm(model, data)

  try {
    const response = await instance.patch(`/${model}/${id}`, formData, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function deleteRequest(model: Model, id: string | number) {
  try {
    const response = await instance.delete(`/${model}/${id}`, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}
