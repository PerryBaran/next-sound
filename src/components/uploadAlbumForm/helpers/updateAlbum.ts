import { AppRouterInstance } from "next/dist/shared/lib/app-router-context"
import { patchAlbum } from "../../../requests/albums"
import { patchSong, deleteSong, postSongs } from "../../../requests/songs"

interface Data {
  name: string
  id: string
  Songs?: {
    name: string
    position: number
    id: string
  }[]
  albumId: string
}

interface Album {
  name: string
  image?: File
}

interface Songs {
  current: {
    name: string
    audio?: File
    delete?: boolean
  }
  original?: {
    name: string
    position: number
    id: string
  }
  key: string
}

export default async function handleSubmit(
  data: Data,
  album: Album,
  songs: Songs[],
  setAlert: React.Dispatch<React.SetStateAction<string>>,
  router: AppRouterInstance,
  name: string
) {
  const albumPromise = []

  if (!album.name) {
    setAlert("Album must have a name")
    return
  }

  if (album.name !== data.name || album.image) {
    albumPromise.push({
      name: album.name !== data.name ? album.name : undefined,
      image: album.image || undefined
    })
  }

  const { length } = songs
  let position = 0
  let songDeletePromises = []
  let songPatchPromises = []
  let songPostPromises = []

  for (let i = 0; i < length; i++) {
    const song = songs[i]
    if (song.original) {
      if (song.current.delete) {
        songDeletePromises.push(song.original.id)
      } else if (!song.current.name) {
        setAlert("All songs must have a name")
        return
      } else if (
        song.current.name !== song.original.name ||
        song.current.audio ||
        song.original.position !== position
      ) {
        const data = {
          name:
            song.current.name !== song.original.name
              ? song.current.name
              : undefined,
          audio: song.current.audio || undefined,
          position: song.original.position !== position ? position : undefined
        }

        songPatchPromises.push({
          id: song.original.id,
          data
        })
      }
    } else {
      if (!song.current.name) {
        setAlert("All songs must have a name")
        return
      }

      if (!song.current.audio) {
        setAlert("All songs must have audio")
        return
      }

      const data = {
        name: song.current.name,
        audio: song.current.audio,
        position
      }

      songPostPromises.push(data)
    }

    if (!song.current.delete) {
      position++
    }
  }

  try {
    await Promise.all([
      ...albumPromise.map((album) => patchAlbum(data.albumId, album)),
      ...songDeletePromises.map((id) => deleteSong(id)),
      ...songPatchPromises.map((song) => patchSong(song.id, song.data)),
      ...songPostPromises.map((song) =>
        postSongs({ ...song, AlbumId: data.albumId })
      )
    ])

    router.push(`profile/${name}`)
  } catch (err: any | Error) {
    setAlert(err?.message || "Unexpected Error")
  }
}
