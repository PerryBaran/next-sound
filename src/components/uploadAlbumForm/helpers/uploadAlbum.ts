import { AppRouterInstance } from "next/dist/shared/lib/app-router-context"
import { postAlbums } from "../../../requests/albums"
import { postSongs } from "../../../requests/songs"

interface Album {
  name: string
  image?: File
}

interface Songs {
  name: string
  audio?: File
}

export default async function uploadAlbum(
  album: Album,
  songs: Songs[],
  setAlert: React.Dispatch<React.SetStateAction<string>>,
  router: AppRouterInstance,
  name: string
) {
  if (!album.name) {
    setAlert("Must provide a name")
    return
  }

  const { length } = songs
  if (length === 0) {
    setAlert("album must contain atleast one song")
    return
  }

  const songPromises = []

  for (let i = 0; i < length; i++) {
    const { name, audio } = songs[i]
    if (!name) {
      setAlert("All songs must contain a name")
      return
    }
    if (!audio) {
      setAlert("All songs must contain audio")
      return
    }
    const data = {
      name,
      audio,
      position: i
    }
    songPromises.push(data)
  }

  try {
    const { id } = await postAlbums(album)
    await Promise.all(
      songPromises.map((song) => {
        return postSongs({ ...song, AlbumId: id })
      })
    )
    router.push(`profile/${name}`)
  } catch (err: any | Error) {
    setAlert(err?.message || "Unexpected Error")
  }
}
