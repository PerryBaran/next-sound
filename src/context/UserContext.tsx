'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
