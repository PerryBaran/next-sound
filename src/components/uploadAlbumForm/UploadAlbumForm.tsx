"use client"

import { useState, useEffect, useRef } from "react"
import { useRouter } from "next/navigation"
import Alert from "@/components/alert/Alert"
import { deleteAlbum } from "@/requests/albums"
import { useUserContext } from "@/context/UserContext"
import css from "./editAlbum.module.css"
import Confirm from "../confirm/Confirm"
import updateAlbum from "./helpers/updateAlbum"
import uploadAlbum from "./helpers/uploadAlbum"

interface Props {
  data?: {
    name: string
    id: string
    Songs?: {
      name: string
      position: number
      id: string
    }[]
    albumId: string
  }
}

interface Album {
  name: string
  image?: File
}

interface Songs {
  current: {
    name: string
    audio?: File
    delete?: boolean
  }
  original?: {
    name: string
    position: number
    id: string
  }
  key: string
}

export default function UploadAlbumForm({ data }: Props) {
  const [album, setAlbum] = useState<Album>({
    name: data ? data.name : ""
  })
  const [songs, setSongs] = useState<Songs[]>(
    data && data.Songs
      ? data.Songs.map((song) => {
          return {
            current: {
              name: song.name,
              delete: false
            },
            original: {
              name: song.name,
              position: song.position,
              id: song.id
            },
            key: crypto.randomUUID()
          }
        })
      : [
          {
            current: {
              name: "",
              delete: false
            },
            key: crypto.randomUUID()
          }
        ]
  )
  const [alert, setAlert] = useState("")
  const dragOver = useRef(0)
  const router = useRouter()
  const {
    user: { name }
  } = useUserContext()
  const [confirm, setConfirm] = useState("")

  const handleAlbumNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target

    setAlbum((prev) => {
      return { ...prev, name: value }
    })
  }

  const handleAlbumImageChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { files } = e.target

    if (!files) return

    setAlbum((prev) => {
      return { ...prev, image: files[0] }
    })
  }

  const handleSongNameChange = (
    e: React.ChangeEvent<HTMLInputElement>,
    i: number
  ) => {
    const { value } = e.target

    setSongs((prev) => {
      const clone = [...prev]
      clone[i].current.name = value
      return clone
    })
  }

  const handleSongAudioChange = (
    e: React.ChangeEvent<HTMLInputElement>,
    i: number
  ) => {
    const { files } = e.target

    if (!files) return

    setSongs((prev) => {
      const clone = [...prev]
      clone[i].current.audio = files[0]
      return clone
    })
  }

  const handleSongDelete = (
    e: React.ChangeEvent<HTMLInputElement>,
    i: number
  ) => {
    const { checked } = e.target

    setSongs((prev) => {
      const clone = [...prev]
      clone[i].current.delete = checked
      return clone
    })
  }

  const dragEnter = (position: number) => {
    dragOver.current = position
  }

  const dragEnd = (startPosition: number) => {
    const endPosition = dragOver.current

    const _songs = [...songs]
    const songContent = _songs[startPosition]
    _songs.splice(startPosition, 1)
    _songs.splice(endPosition, 0, songContent)
    setSongs(_songs)
  }

  const addSong = () => {
    setSongs((prev) => [
      ...prev,
      {
        current: {
          name: ""
        },
        key: crypto.randomUUID()
      }
    ])
  }

  const removeSong = (i: number) => {
    setSongs((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })
  }

  const handleSubmit = async () => {
    if (data) {
      await updateAlbum(data, album, songs, setAlert, router, name)
    } else {
      await uploadAlbum(
        album,
        songs.map((song) => song.current),
        setAlert,
        router,
        name
      )
    }
  }

  const handleDeleteAlbum = async () => {
    if (data) {
      try {
        await deleteAlbum(data.albumId)

        router.push(`profile/${name}`)
      } catch (err: any | Error) {
        setAlert(err?.message || "Unexpected Error")
      }
    } else {
      setAlert("Cannot delete album")
    }
  }

  const handleCancel = () => {
    router.back()
  }

  const confirmCallback = (string: string) => {
    switch (string) {
      case "cancel":
        return handleCancel
      case "submit":
        return handleSubmit
      case "delete":
        return handleDeleteAlbum
      default:
        return () => {}
    }
  }

  return (
    <div>
      <Alert message={alert} />
      <form
        onSubmit={(e: React.FormEvent) => {
          e.preventDefault()
          setConfirm("submit")
        }}
      >
        <h2>{data ? "Edit Album" : "Upload Album"}</h2>
        <div>
          <label htmlFor="album-name">
            <span>Name</span>
            <input
              type="text"
              id="album-name"
              value={album.name}
              onChange={handleAlbumNameChange}
            />
          </label>
          <label htmlFor="art">
            <span>Cover Art</span>
            <input type="file" id="art" onChange={handleAlbumImageChange} />
          </label>
        </div>
        <div>
          <h3>Edit Songs</h3>
          {songs.map((song, i) => {
            return (
              <div
                key={song.key}
                draggable
                onDragEnter={() => dragEnter(i)}
                onDragEnd={() => dragEnd(i)}
                className={css.song}
              >
                <h4>Song {i + 1}</h4>
                <label htmlFor={`song-name${i}`}>
                  <span>Name</span>
                  <input
                    type="text"
                    id={`song-name${i}`}
                    name="name"
                    value={song.current.name}
                    onChange={(e) => handleSongNameChange(e, i)}
                  />
                </label>
                <label htmlFor={`song-audio${i}`}>
                  <span className="upload-info">Audio</span>
                  <input
                    type="file"
                    id={`song-audio${i}`}
                    onChange={(e) => handleSongAudioChange(e, i)}
                  />
                </label>
                {!song.original ? (
                  <button
                    type="button"
                    onClick={() => removeSong(i)}
                    className={css["delete-song"]}
                  >
                    x
                  </button>
                ) : (
                  <label
                    htmlFor={`delete-song${i}`}
                    className={css["delete-song"]}
                  >
                    <span className="upload-info">x</span>
                    <input
                      type="checkbox"
                      id={`delete-song${i}`}
                      name="delete"
                      checked={song.current.delete}
                      onChange={(e) => handleSongDelete(e, i)}
                    />
                  </label>
                )}
              </div>
            )
          })}
          <button type="button" onClick={addSong}>
            +
          </button>
        </div>
        <button type="submit">{data ? "Save Changes" : "Upload"}</button>
        {data && (
          <button type="button" onClick={() => setConfirm("delete")}>
            Delete Album
          </button>
        )}
        <button type="button" onClick={() => setConfirm("cancel")}>
          Cancel
        </button>
        {confirm && (
          <Confirm
            setConfirm={setConfirm}
            callback={confirmCallback(confirm)}
          />
        )}
      </form>
    </div>
  )
}
