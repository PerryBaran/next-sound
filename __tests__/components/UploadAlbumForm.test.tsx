import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import { createContext, useContext } from "react"
import UploadAlbumForm from "../../src/components/uploadAlbumForm/UploadAlbumForm"
import * as navigation from "next/navigation"
import * as UserContext from "../../src/context/UserContext"
import * as updateAlbum from "../../src/components/uploadAlbumForm/helpers/updateAlbum"
import * as uploadAlbum from "../../src/components/uploadAlbumForm/helpers/uploadAlbum"
import * as albumRequests from "../../src/requests/albums"

describe("UploadAlbumForm", () => {
  let mockRouter = {
    back: jest.fn(),
    forward: jest.fn(),
    push: jest.fn(),
    refresh: jest.fn(),
    replace: jest.fn(),
    prefetch: jest.fn()
  }
  const mockName = "mock-name"
  let updateAlbumSpy: jest.SpyInstance
  let uploadAlbumSpy: jest.SpyInstance

  afterEach(() => {
    jest.clearAllMocks()
  })

  beforeEach(() => {
    mockRouter = {
      back: jest.fn(),
      forward: jest.fn(),
      push: jest.fn(),
      refresh: jest.fn(),
      replace: jest.fn(),
      prefetch: jest.fn()
    }

    jest.spyOn(navigation, "useRouter").mockReturnValue(mockRouter)

    const mockContext = createContext({
      user: { name: mockName, id: "" },
      handleLogin: jest.fn()
    })
    jest
      .spyOn(UserContext, "useUserContext")
      .mockImplementation(() => useContext(mockContext))

    updateAlbumSpy = jest.spyOn(updateAlbum, "default").mockResolvedValue()
    uploadAlbumSpy = jest.spyOn(uploadAlbum, "default").mockResolvedValue()
  })

  describe("upload - no data", () => {
    describe("tests", () => {
      beforeEach(() => {
        render(<UploadAlbumForm />)
      })

      test("renders correctly", () => {
        const nameInputs = screen.getAllByLabelText(/name/i)

        expect(screen.getByText(/upload album/i)).toBeInstanceOf(
          HTMLHeadingElement
        )
        expect(nameInputs).toHaveLength(2)
        nameInputs.forEach((input) => {
          expect(input).toHaveAttribute("type", "text")
        })
        expect(screen.getByLabelText(/cover art/i)).toHaveAttribute(
          "type",
          "file"
        )
        expect(screen.getByText(/upload songs/i)).toBeInstanceOf(
          HTMLHeadingElement
        )
        expect(screen.getByText(/song 1/i)).toBeInstanceOf(HTMLHeadingElement)
        expect(screen.getByLabelText(/audio/i)).toHaveAttribute("type", "file")
        expect(screen.getByText(/x/i)).toBeInstanceOf(HTMLButtonElement)
        expect(screen.getByText("+")).toBeInstanceOf(HTMLButtonElement)
        expect(screen.getByText(/submit/i)).toHaveAttribute("type", "submit")
        expect(screen.getByText(/cancel/i)).toBeInstanceOf(HTMLButtonElement)
      })

      test("remove button", () => {
        expect(screen.getAllByLabelText(/name/i)).toHaveLength(2)

        fireEvent.click(screen.getByText(/x/i))

        expect(screen.getAllByLabelText(/name/i)).toHaveLength(1)
      })

      test("add button", () => {
        expect(screen.getAllByLabelText(/audio/i)).toHaveLength(1)

        fireEvent.click(screen.getByText("+"))

        expect(screen.getAllByLabelText(/audio/i)).toHaveLength(2)
      })

      test("drag and drop", () => {
        fireEvent.click(screen.getByText("+"))

        const name = "name"
        fireEvent.change(screen.getAllByLabelText(/name/i)[1], {
          target: { value: name }
        })

        fireEvent.dragEnter(screen.getAllByLabelText(/name/i)[2])
        fireEvent.dragEnd(screen.getAllByLabelText(/name/i)[1])

        expect(screen.getAllByLabelText(/name/i)[2]).toHaveValue(name)
      })

      test("submit", () => {
        fireEvent.click(screen.getByText(/submit/i))

        expect(uploadAlbumSpy).toBeCalledTimes(0)

        fireEvent.click(screen.getByText(/yes/i))

        expect(uploadAlbumSpy).toBeCalledTimes(1)
        expect(updateAlbumSpy).toBeCalledTimes(0)
      })

      test("cancel", () => {
        fireEvent.click(screen.getByText(/cancel/i))

        expect(mockRouter.back).toBeCalledTimes(0)

        fireEvent.click(screen.getByText(/yes/i))

        expect(mockRouter.back).toBeCalledTimes(1)
      })
    })

    test("snapshot", () => {
      const { asFragment } = render(<UploadAlbumForm />)

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe("update - with data", () => {
    const mockData = {
      name: "album-name",
      id: "album-id",
      Songs: [
        {
          name: "song-1-name",
          position: 0,
          id: "song-1-id"
        },
        {
          name: "song-2-name",
          position: 1,
          id: "song-2-id"
        }
      ]
    }

    describe("tests", () => {
      beforeEach(() => {
        render(<UploadAlbumForm data={mockData} />)
      })

      test("renders correctly", () => {
        const nameInputs = screen.getAllByLabelText(/name/i)

        expect(screen.getByText(/edit album/i)).toBeInstanceOf(
          HTMLHeadingElement
        )
        expect(nameInputs).toHaveLength(3)
        nameInputs.forEach((input, i) => {
          if (i === 0) {
            expect(input).toHaveValue(mockData.name)
          } else {
            expect(input).toHaveValue(mockData.Songs[i - 1].name)
          }
        })
        expect(screen.getByLabelText(/cover art/i)).toHaveAttribute(
          "type",
          "file"
        )
        expect(screen.getByText(/edit songs/i)).toBeInstanceOf(
          HTMLHeadingElement
        )
        expect(screen.getByText(/song 1/i)).toBeInstanceOf(HTMLHeadingElement)
        expect(screen.getAllByLabelText(/audio/i)[0]).toHaveAttribute(
          "type",
          "file"
        )
        expect(screen.getAllByLabelText(/x/i)[0]).toBeInstanceOf(
          HTMLInputElement
        )
        expect(screen.getByText("+")).toBeInstanceOf(HTMLButtonElement)
        expect(screen.getByText(/save changes/i)).toHaveAttribute(
          "type",
          "submit"
        )
        expect(screen.getByText(/delete album/i)).toBeInstanceOf(
          HTMLButtonElement
        )
        expect(screen.getByText(/cancel/i)).toBeInstanceOf(HTMLButtonElement)
      })

      test("remove input", () => {
        const remove = screen.getAllByLabelText(/x/i)[0]

        expect(remove).not.toBeChecked()

        fireEvent.click(remove)

        expect(remove).toBeChecked()
      })

      test("add button", () => {
        expect(screen.getAllByLabelText(/audio/i)).toHaveLength(2)

        fireEvent.click(screen.getByText("+"))

        expect(screen.getAllByLabelText(/audio/i)).toHaveLength(3)
      })

      test("remove button", () => {
        expect(screen.getAllByLabelText(/audio/i)).toHaveLength(2)

        fireEvent.click(screen.getByText("+"))

        expect(screen.getAllByLabelText(/audio/i)).toHaveLength(3)

        const removeButtons = screen.getAllByText(/x/i)
        fireEvent.click(removeButtons[removeButtons.length - 1])

        expect(screen.getAllByLabelText(/audio/i)).toHaveLength(2)
      })

      test("drag and drop", () => {
        const names = screen.getAllByLabelText(/name/i)

        expect(names[1]).toHaveValue(mockData.Songs[0].name)

        fireEvent.dragEnter(names[2])
        fireEvent.dragEnd(names[1])

        expect(screen.getAllByLabelText(/name/i)[2]).toHaveValue(
          mockData.Songs[0].name
        )
      })

      test("submit", () => {
        fireEvent.click(screen.getByText(/save changes/i))

        expect(updateAlbumSpy).toBeCalledTimes(0)

        fireEvent.click(screen.getByText(/yes/i))

        expect(updateAlbumSpy).toBeCalledTimes(1)
        expect(uploadAlbumSpy).toBeCalledTimes(0)
      })

      test("cancel", () => {
        fireEvent.click(screen.getByText(/cancel/i))

        expect(mockRouter.back).toBeCalledTimes(0)

        fireEvent.click(screen.getByText(/yes/i))

        expect(mockRouter.back).toBeCalledTimes(1)
      })

      describe("delete", () => {
        let deleteAlbumSpy: jest.SpyInstance
        beforeEach(() => {
          deleteAlbumSpy = jest.spyOn(albumRequests, "deleteAlbum")
        })

        test("succesfully deletes", async () => {
          deleteAlbumSpy.mockResolvedValue("")
          fireEvent.click(screen.getByText(/delete/i))

          expect(mockRouter.push).toBeCalledTimes(0)
          expect(deleteAlbumSpy).toBeCalledTimes(0)

          fireEvent.click(screen.getByText(/yes/i))

          await waitFor(() => {
            expect(deleteAlbumSpy).toBeCalledWith(mockData.id)
            expect(mockRouter.push).toBeCalledWith(`profile/${mockName}`)
          })
        })

        test("throws expected error", async () => {
          const error = "error"
          deleteAlbumSpy.mockRejectedValue({ message: error })

          fireEvent.click(screen.getByText(/delete/i))

          fireEvent.click(screen.getByText(/yes/i))

          await waitFor(() => {
            expect(screen.getByText(error)).toBeInTheDocument()
          })
        })

        test("throws unexpected error", async () => {
          deleteAlbumSpy.mockRejectedValue("")

          fireEvent.click(screen.getByText(/delete/i))

          fireEvent.click(screen.getByText(/yes/i))

          await waitFor(() => {
            expect(screen.getByText("Unexpected Error")).toBeInTheDocument()
          })
        })
      })
    })

    test("snapshot", () => {
      const { asFragment } = render(<UploadAlbumForm data={mockData} />)

      expect(asFragment()).toMatchSnapshot()
    })
  })
})
