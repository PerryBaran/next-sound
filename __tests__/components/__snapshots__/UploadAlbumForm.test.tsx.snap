// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadAlbumForm update - with data snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <h2>
        Edit Album
      </h2>
      <div>
        <label
          for="album-name"
        >
          <span>
            Name
          </span>
          <input
            id="album-name"
            type="text"
            value="album-name"
          />
        </label>
        <label
          for="art"
        >
          <span>
            Cover Art
          </span>
          <input
            id="art"
            type="file"
          />
        </label>
      </div>
      <div>
        <h3>
          Edit Songs
        </h3>
        <div
          class="song"
          draggable="true"
        >
          <h4>
            Song 1
          </h4>
          <label
            for="song-name0"
          >
            <span>
              Name
            </span>
            <input
              id="song-name0"
              name="name"
              type="text"
              value="song-1-name"
            />
          </label>
          <label
            for="song-audio0"
          >
            <span
              class="upload-info"
            >
              Audio
            </span>
            <input
              id="song-audio0"
              type="file"
            />
          </label>
          <label
            class="delete-song"
            for="delete-song0"
          >
            <span
              class="upload-info"
            >
              x
            </span>
            <input
              id="delete-song0"
              name="delete"
              type="checkbox"
            />
          </label>
        </div>
        <div
          class="song"
          draggable="true"
        >
          <h4>
            Song 2
          </h4>
          <label
            for="song-name1"
          >
            <span>
              Name
            </span>
            <input
              id="song-name1"
              name="name"
              type="text"
              value="song-2-name"
            />
          </label>
          <label
            for="song-audio1"
          >
            <span
              class="upload-info"
            >
              Audio
            </span>
            <input
              id="song-audio1"
              type="file"
            />
          </label>
          <label
            class="delete-song"
            for="delete-song1"
          >
            <span
              class="upload-info"
            >
              x
            </span>
            <input
              id="delete-song1"
              name="delete"
              type="checkbox"
            />
          </label>
        </div>
        <button
          type="button"
        >
          +
        </button>
      </div>
      <button
        type="submit"
      >
        Save Changes
      </button>
      <button
        type="button"
      >
        Delete Album
      </button>
      <button
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`UploadAlbumForm upload - no data snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <h2>
        Upload Album
      </h2>
      <div>
        <label
          for="album-name"
        >
          <span>
            Name
          </span>
          <input
            id="album-name"
            type="text"
            value=""
          />
        </label>
        <label
          for="art"
        >
          <span>
            Cover Art
          </span>
          <input
            id="art"
            type="file"
          />
        </label>
      </div>
      <div>
        <h3>
          Upload Songs
        </h3>
        <div
          class="song"
          draggable="true"
        >
          <h4>
            Song 1
          </h4>
          <label
            for="song-name0"
          >
            <span>
              Name
            </span>
            <input
              id="song-name0"
              name="name"
              type="text"
              value=""
            />
          </label>
          <label
            for="song-audio0"
          >
            <span
              class="upload-info"
            >
              Audio
            </span>
            <input
              id="song-audio0"
              type="file"
            />
          </label>
          <button
            class="delete-song"
            type="button"
          >
            x
          </button>
        </div>
        <button
          type="button"
        >
          +
        </button>
      </div>
      <button
        type="submit"
      >
        Submit
      </button>
      <button
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;
